x-pack/legacy/plugins/reporting/server/plugin.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Legacy } from 'kibana';
+import { CoreSetup, CoreStart, Plugin } from 'src/core/server';
+import { IUiSettingsClient } from 'src/core/server';
+import { UsageCollectionSetup } from 'src/plugins/usage_collection/server';
+import { XPackMainPlugin } from '../../xpack_main/server/xpack_main';
+// @ts-ignore
+import { mirrorPluginStatus } from '../../../server/lib/mirror_plugin_status';
+import { PLUGIN_ID } from '../common/constants';
+import { ReportingPluginSpecOptions } from '../types.d';
+import { registerRoutes } from './routes';
+import { LevelLogger, checkLicenseFactory, getExportTypesRegistry, runValidations } from './lib';
+import { createBrowserDriverFactory } from './browsers';
+import { registerReportingUsageCollector } from './usage';
+import { logConfiguration } from '../log_configuration';
+
+// For now there is no exposed functionality to other plugins
+export type ReportingSetup = object;
+export type ReportingStart = object;
+
+export interface ReportingSetupDeps {
+  usageCollection: UsageCollectionSetup;
+}
+export type ReportingStartDeps = object;
+
+type LegacyPlugins = Legacy.Server['plugins'];
+
+export interface LegacySetup {
+  config: Legacy.Server['config'];
+  info: Legacy.Server['info'];
+  log: Legacy.Server['log'];
+  plugins: {
+    elasticsearch: LegacyPlugins['elasticsearch'];
+    security: LegacyPlugins['security'];
+    xpack_main: XPackMainPlugin & {
+      status?: any;
+    };
+  };
+  route: Legacy.Server['route'];
+  savedObjects: Legacy.Server['savedObjects'];
+  uiSettingsServiceFactory: Legacy.Server['uiSettingsServiceFactory'];
+  fieldFormatServiceFactory: (uiConfig: IUiSettingsClient) => unknown;
+}
+
+export type ReportingPlugin = Plugin<
+  ReportingSetup,
+  ReportingStart,
+  ReportingSetupDeps,
+  ReportingStartDeps
+>;
+
+/* We need a factory that returns an instance of the class because the class
+ * implementation itself restricts against having Legacy dependencies passed
+ * into `setup`. The factory parameters take the legacy dependencies, and the
+ * `setup` method gets it from enclosure */
+export function reportingPluginFactory(
```

Comment:
I made this factory function to accept parameters that can not directly be passed into the `ReportingPlugin` constructor and setup method.

I had to come up with this solution, because the guidance from the [Migration guide](https://github.com/elastic/kibana/blob/master/src/core/MIGRATION.md#introduce-new-plugin-definition-shim) had outdated info:

```
const demoSetup = new Plugin().setup(coreSetup, pluginsSetup, __LEGACY); // this is not possible if the Plugin class extends the core plugin class
```

Reply:
We could make this type more permissive:

```ts
interface Plugin<...> {
  setup(core: CoreSetup, plugins: TPluginsSetup, ...args: any[]): MaybePromise<TSetup>;
  /* ... */
}
```

Reply:
(sorry, I'm a little late) I'm not a big fan of 'poluting' the `Plugin` interface, as the additional parameters will not be able to actually be injected/used once the plugin is migrated and `core` actually instantiate the plugin.

Maybe we could instead introduce a sub-interface for the migration period,

```typescript
interface ShimedPlugin<
  TSetup = void,
  TStart = void,
  TPluginsSetup extends object = object,
  TPluginsStart extends object = object,
  TLegacySetupDeps extends object = object
> extends Plugin<TSetup, TStart, TPluginsSetup, TPluginsStart> {
  setup(
    core: CoreSetup,
    plugins: TPluginsSetup,
    legacy: TLegacySetupDeps
  ): TSetup | Promise<TSetup>;
}
```

@tsullivan The factory is a workaround that will do just fine however. Sorry for the outdated documentation